R/programming.R

```diff
@@ -0,0 +1,31 @@
+substitute2 = function(expr, env, char.as.name=FALSE, sub.names=TRUE) {
+  if (missing(env)) {
+    stop("TODO")
+  } else if (is.environment(env)) {
+    env = as.list(env, all.names=TRUE) ## todo: try to use environment rather than list
+  } else if (!is.list(env)) {
+    stop("'env' must be a list of an environment")
+  }
+  env.names = names(env)
+  if (is.null(env.names)) {
+    stop("'env' argument does not have names")
+  } else if (!all(nzchar(env.names))) {
+    stop("'env' argument has an zero char names")
+  }
+  if (isTRUE(char.as.name)) {
+    char = vapply(env, is.character, FALSE)
+    if (any(char)) {
+      if (any(non.scalar.char <- lengths(env[char])!=1L)) {
```

Comment:
`lengths` is not 3.1.0 right? Maybe we should just backport it like we did `trimws`

Reply:
Not big issue about vapply instead, but what I would like to have more is to coerce 'env' to environment and use that rather than a list. This is how base R uses 'env' argument in 'substitute'. Anyway need to measure performance before deciding which one to use.